pkg/resource/defaultclass.go

```diff
@@ -0,0 +1,133 @@
+type DefaultClassReconciler struct {
+	client   client.Client
+	newClaim func() Claim
+	options  *client.ListOptions
+}
+
+// NewDefaultClassReconciler creates a new DefaultReconciler for the claim kind
+func NewDefaultClassReconciler(m manager.Manager, of ClaimKind) *DefaultClassReconciler {
+	nc := func() Claim { return MustCreateObject(schema.GroupVersionKind(of), m.GetScheme()).(Claim) }
+
+	// Panic early if we've been asked to reconcile a claim that has
+	// not been registered with our controller manager's scheme.
+	_ = nc()
+
+	gk := strings.ToLower(schema.GroupVersionKind(of).GroupKind().String())
+
+	// Create list options query that will be used to search
+	// for resource class that is default for claim kind.
+	options := &client.ListOptions{}
+	if err := options.SetLabelSelector(gk + "/default=" + "true"); err != nil {
+		// Panic if unable to set label selector or else panic will occur
+		// when returned reconciler is invoked.
+		panic(err)
+	}
+
+	return &DefaultClassReconciler{
+		client:   m.GetClient(),
+		newClaim: nc,
+		options:  options,
+	}
+}
+
+// Reconcile reconciles a claim to the default class reference for its kind
+func (r *DefaultClassReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	logDefaultClass.V(logging.Debug).Info("Reconciling", "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	claim := r.newClaim()
+	if err := r.client.Get(ctx, req.NamespacedName, claim); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetClaim)
+	}
+
+	// Get resource classes with claim kind as default
+	classes := &corev1alpha1.ResourceClassList{}
+	if err := r.client.List(ctx, r.options, classes); err != nil {
+		// If this is the first time we encounter no defaults we'll be
+		// requeued implicitly due to the status update. If not, we don't
+		// care to requeue because list parameters will not change.
+		claim.SetConditions(corev1alpha1.ReconcileError(errors.New(errFailedList)))
+		return reconcile.Result{}, errors.Wrap(IgnoreNotFound(r.client.Status().Update(ctx, claim)), errUpdateClaimStatus)
+	}
+
+	// Check to see if no defaults defined for claim kind.
+	if len(classes.Items) == 0 {
+		// If this is the first time we encounter no defaults we'll be
+		// requeued implicitly due to the status update. If not, we will requeue
+		// after a time to see if a default class has been created.
+		claim.SetConditions(corev1alpha1.ReconcileError(errors.New(errNoDefaultClass)))
+		return reconcile.Result{RequeueAfter: defautlClassWait}, errors.Wrap(IgnoreNotFound(r.client.Status().Update(ctx, claim)), errUpdateClaimStatus)
+	}
+
+	// Check to see if multiple defaults defined for claim kind.
+	if len(classes.Items) > 1 {
+		// If this is the first time we encounter no defaults we'll be
```

Comment:
"Too many defaults" rather than "no defaults"?

We could merge this with the "no default class" case by having a test an error for "exactly one one default resource class did not exist". I'm ambivalent about whether we should.

Reply:
Yep! This comment should say "too many defaults", good catch!

I am not sure what you are saying in the second part of your comment, @negz could you clarify for me?

Reply:
> @negz could you clarify for me?

I was thinking you could have a single test for `len(classes.Items) != 1` rather than checking `len(classes.Items) == 0` and also `len(classes.Items) > 1`. The advantage is simpler code, but the disadvantage a vague error message (e.g. "exactly one default resource class must exist" rather than "no default resource class" exists.

As I write this, I lean toward your current implementation being better.